src/sinks/gcp_pubsub.rs

```diff
@@ -0,0 +1,429 @@
+                    Ok(())
+                }
+                status => Err(super::HealthcheckError::UnexpectedStatus { status }.into()),
+            });
+
+        Ok(Box::new(healthcheck))
+    }
+
+    fn uri(&self, suffix: &str) -> crate::Result<Uri> {
+        let base = match self.emulator_host.as_ref() {
+            Some(host) => format!("http://{}", host),
+            None => "https://pubsub.googleapis.com".into(),
+        };
+        let uri = format!(
+            "{}/v1/projects/{}/topics/{}{}",
+            base, self.project, self.topic, suffix
+        );
+        let uri = match &self.api_key {
+            Some(key) => format!("{}?key={}", uri, key),
+            None => uri,
+        };
+        uri.parse::<Uri>()
+            .context(super::UriParseError)
+            .map_err(Into::into)
+    }
+}
+
+#[derive(Clone)]
+struct PubsubCreds {
+    creds: Credentials,
+    token: Arc<RwLock<Token>>,
+}
+
+impl PubsubCreds {
+    fn new(path: &str) -> crate::Result<Self> {
+        let creds = Credentials::from_file(path).context(InvalidCredentials)?;
+        let jwt = make_jwt(&creds)?;
+        let token = goauth::get_token_with_creds(&jwt, &creds).context(GetTokenFailed)?;
+        let token = Arc::new(RwLock::new(token));
+        Ok(Self { creds, token })
+    }
+
+    fn apply<T>(&self, request: &mut Request<T>) {
+        let token = self.token.read().unwrap();
+        let value = format!("{} {}", token.token_type(), token.access_token());
+        request
+            .headers_mut()
+            .insert(AUTHORIZATION, HeaderValue::from_str(&value).unwrap());
+    }
+
+    fn regenerate_token(&self) -> crate::Result<()> {
+        let jwt = make_jwt(&self.creds).unwrap(); // Errors caught above
+        let token = goauth::get_token_with_creds(&jwt, &self.creds)?;
+        *self.token.write().unwrap() = token;
+        Ok(())
+    }
+
+    fn spawn_regenerate_token(&self) {
+        let interval = self.token.read().unwrap().expires_in() as u64 / 2;
+        let copy = self.clone();
+        let renew_task = Interval::new_interval(Duration::from_secs(interval))
+            .for_each(move |_instant| {
+                debug!("Renewing GCP pubsub token");
+                if let Err(error) = copy.regenerate_token() {
+                    error!(message = "Failed to update GCP pubsub token", %error);
+                }
+                Ok(())
+            })
+            .map_err(
+                |error| error!(message = "GCP pubsub token regenerate interval failed", %error),
+            );
+
+        tokio::spawn(renew_task);
+    }
+}
+
+fn make_jwt(creds: &Credentials) -> crate::Result<Jwt<JwtClaims>> {
+    let claims = JwtClaims::new(creds.iss(), &Scope::PubSub, creds.token_uri(), None, None);
+    let rsa_key = creds.rsa_key().context(InvalidRsaKey)?;
+    Ok(Jwt::new(claims, rsa_key, None))
+}
+
+fn make_body(logs: Vec<u8>) -> Vec<u8> {
+    let mut body = BytesMut::with_capacity(logs.len() + 16);
```

Comment:
Could usage of the `16` literal be avoided to prevent forgetting to update it in case if the header and footer need to be changed? For example, named constants `header` and `footer` for `"{\"messages\":["` and `"]}"` could have been introduced and then their lengths used instead.

Reply:
Absolutely. Will do.